File: common/src/main/java/dev/obscuria/fragmentum/core/CoreFragmentum.java
package dev.obscuria.fragmentum.core;

import dev.obscuria.fragmentum.api.v1.common.FragmentumEvents;
import dev.obscuria.fragmentum.api.v1.common.text.AutoDescription;
import dev.obscuria.fragmentum.api.v1.common.text.TextWrapper;
import dev.obscuria.fragmentum.core.v1.common.AllOfTooltipComponent;
import net.minecraft.core.registries.BuiltInRegistries;
import net.minecraft.network.chat.Component;
import net.minecraft.network.chat.Style;
import net.minecraft.world.inventory.tooltip.TooltipComponent;
import net.minecraft.world.item.ItemStack;
import org.apache.commons.compress.utils.Lists;
import org.jetbrains.annotations.ApiStatus;

import java.util.Optional;
import java.util.function.Consumer;

@ApiStatus.Internal
public final class CoreFragmentum
{
    public static void init() {}

    public static void modifyTooltip(ItemStack stack, Consumer<Component> consumer)
    {
        if (!stack.getItem().getClass().isAnnotationPresent(AutoDescription.class)) return;
        final var annotation = stack.getItem().getClass().getAnnotation(AutoDescription.class);
        final var id = BuiltInRegistries.ITEM.getKey(stack.getItem());
        final var description = Component.translatable(id.toLanguageKey("item", "desc"));
        final var lines = TextWrapper.create(description)
                .withStyle(Style.EMPTY.withColor(annotation.style()))
                .withMaxLength(40)
                .build();
        lines.forEach(consumer);
    }

    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
    public static Optional<TooltipComponent> gatherTooltipImages(ItemStack stack,
                                                                 Optional<TooltipComponent> defaultImage)
    {
        final var customImages = Lists.<TooltipComponent>newArrayList();
        FragmentumEvents.GATHER_TOOLTIP_IMAGES.broadcast(listener -> listener.invoke(stack, customImages::add));

        if (customImages.isEmpty()) return defaultImage;
        defaultImage.ifPresent(customImages::addFirst);
        return Optional.of(new AllOfTooltipComponent(customImages));
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/core/CoreFragmentum.kt
package dev.obscuria.fragmentum.core

import dev.obscuria.fragmentum.api.FragmentumEvents
import dev.obscuria.fragmentum.api.tools.text.AutoDescription
import dev.obscuria.fragmentum.api.tools.text.TextWrapper
import dev.obscuria.fragmentum.core.world.AllOfTooltipComponent
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.network.chat.Component
import net.minecraft.network.chat.Style
import net.minecraft.world.inventory.tooltip.TooltipComponent
import net.minecraft.world.item.ItemStack
import org.apache.commons.compress.utils.Lists
import org.jetbrains.annotations.ApiStatus.Internal
import java.util.*
import java.util.function.Consumer

@Internal
object CoreFragmentum
{
    fun init()
    {
    }

    @JvmStatic
    fun modifyTooltip(stack: ItemStack, consumer: Consumer<Component>)
    {
        if (!stack.item.javaClass.isAnnotationPresent(AutoDescription::class.java)) return
        val annotation = stack.item.javaClass.getAnnotation(AutoDescription::class.java)
        val id = BuiltInRegistries.ITEM.getKey(stack.item)
        val description = Component.translatable(id.toLanguageKey("item", "desc"))
        val lines = TextWrapper.create(description)
            .withStyle(Style.EMPTY.withColor(annotation.style))
            .withMaxLength(40)
            .build()
        lines.forEach(consumer)
    }

    @JvmStatic
    fun gatherTooltipImages(stack: ItemStack, defaultImage: Optional<TooltipComponent>): Optional<TooltipComponent>
    {
        val customImages = Lists.newArrayList<TooltipComponent>()
        FragmentumEvents.GATHER_TOOLTIP_IMAGES broadcast { listener -> listener.invoke(stack, customImages::add) }
        if (customImages.isEmpty()) return defaultImage

        defaultImage.ifPresent { image: TooltipComponent -> customImages.add(0, image) }
        return Optional.of<TooltipComponent>(AllOfTooltipComponent(customImages))
    }
}